Explore Multi-Entity Saving

epiqueras opened the pull request:
Follows #17368

## Description

This PR explores what it could look like to save multiple entities at the same time using a modal with a list of checkboxes that let you select a set of entity changes to save.

## How has this been tested?

It was verified that:

- The new button appears only when there are pending changes to an entity and the full site editing experiment is enabled.
- Clicking the button opens a modal with a checkbox list of entities with pending changes and their changed properties.
- Selecting a set of entities and clicking the save button works as expected, only saving the selected entities.
- If there are no selected entities, the save button is disabled.

## Screenshots

<img width="846" alt="Screen Shot 2019-10-18 at 1 23 01 PM" src="https://user-images.githubusercontent.com/19157096/67125869-94852980-f1aa-11e9-97aa-880fc94039f9.png">

## Types of Changes

*New Feature:* The full site editing experiment now has a button for selecting and saving multiple entities at once.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

epiqueras committed:
Core Data: Add selector for getting all entity changes.

epiqueras committed:
Editor: Add global changes save button.

jasmussen commented:
Nice work. I don't have time to review in depth today, will test this out tomorrow. However based on looking just at the screenshots, here are some high-level thoughts:

1. We need to not have a separate "Save global edits" button. 
2. If only ever one thing changed, we should not need a prompt.

Those are principles we should aim for, but obviously there's a lot of technical reasoning to work out. For example, in the screenshot, what is actualy being saved? A draft? Would pressing Save on the blue button be the same as pressing "Save Draft"?

Tricky use case: user is working on a draft and is not ready to publish it. But they also made changes to the Header block area. What happens?

A few thoughts to help us work through it:

- I still believe that "save draft" is not something a user should ever have to think of. It should happen automatically and there should not be an explicit save button unless something went wrong. This is the Figma auto-save model. 
- If a user is working on a draft and made global edits, perhaps we should hide the "Save Draft" button until the global edits are published. 
- Given global edits are being _published_, perhaps the actual Publish button should be used. 

<img width="637" alt="Screenshot 2019-10-21 at 11 56 04" src="https://user-images.githubusercontent.com/1204802/67195791-e8398200-f3f9-11e9-880c-737181092bb9.png">

That's a screenshot from https://github.com/WordPress/gutenberg/issues/13489#issuecomment-539949749. In it, global changes were made, so there's not a "Save Draft" button, instead there's a "There are unpublished changes" label. The way you publish those is using the Publish button, which then opens the modal. 

But perhaps that label needs a little love. Here's a tweaked flow:

<img width="2830" alt="Publishing Block Area Changes" src="https://user-images.githubusercontent.com/1204802/67196093-6eee5f00-f3fa-11e9-810e-476dfb28c03d.png">

Note in the above:

- How there's no save draft button.
- That there's a Publish Changes label.
- That the Page itself, which is a draft, is unchecked by default.

What would happen if you pressed "Publish" in the above screenshot, is that you would:

- Publish the header change
- Get the Save Draft button back, because now you're just editing a draft
- The Publish button would change to say "Publish".

What do you think?

Also, CC: @shaunandrews because I know you have thoughts here as well.

epiqueras commented:
>We need to not have a separate "Save global edits" button.

Yes, but some sort of separation like this would make things easier to grasp.

>If only ever one thing changed, we should not need a prompt.

What if that one thing is not the actual post you are editing. It'd be confusing if you saved as usual.

>I still believe that "save draft" is not something a user should ever have to think of. It should happen automatically and there should not be an explicit save button unless something went wrong. This is the Figma auto-save model.

I agree, and this would simplify our problem here. We would no longer have to think about what the save draft button does in relation to global edits.

>If a user is working on a draft and made global edits, perhaps we should hide the "Save Draft" button until the global edits are published.

What if users want to save their post content, but not the other changes yet.

>Given global edits are being published, perhaps the actual Publish button should be used.

This creates confusion between publishing global changes in a draft and actually publishing the post.

>But perhaps that label needs a little love. Here's a tweaked flow:

This is nice, but a bit confusing. What if I want to save the draft without publishing the global changes? I would have to manually undo all the global changes? How does "Publish Changes" convey to users that they're doing something different to "Publish", what if this is an already published post?

I think that the optimal scenario is to have the auto-save draft like Figma for post content, and then for the global changes, we should either style or iconize the publish button to convey that the modal will open, and this can let them just save or publish just the post content, or we can have another button next to the publish button appear, something along the lines of "Publish Templates".

I think the styling approach might be a bit more elegant. What do you think?

jasmussen commented:
You make some good points, all worth addressing. 

By not adding additional buttons, I mean to the toolbar. There literally isn't space:

![Screenshot 2019-10-22 at 10 20 07](https://user-images.githubusercontent.com/1204802/67268746-8b030680-f4b5-11e9-947b-350cfa06c095.png)

If we absolutely have to have a separate button to publish changes made to template parts, we have to explore other designs, such as changing the publish button contextually, or making it a split button or something else. But I don't yet fully accept that this is necessary.

> Yes, but some sort of separation like this would make things easier to grasp.

This is a balancing act that has to consider how often you make global edits compared to how often you just write a post or a page. One extra button discussed in isolation may seem like a small thing, but it all adds up. 

Especially with something as complex as being able to edit MULTIPLE content areas at once, it is crucial that the flow is simple. 

By not having an extra button, but instead bundling multiple changes with the Publish flow, we can keep the writing flow simple, yet give both context, information, and control to the user during the publishing act.

> This is nice, but a bit confusing. What if I want to save the draft without publishing the global changes? I would have to manually undo all the global changes? How does "Publish Changes" convey to users that they're doing something different to "Publish", what if this is an already published post?

Maybe you're right — maybe we don't touch the save draft button, and all it does is save the post content. No new label indicating unsaved global changes, no extra buttons. But maybe we still change the _label_ of the Publish button, and then show the popup. We might even show an "unread dot" or something, overlaid on the Publish button. Like this maybe?

<img width="2830" alt="Publishing Block Area Changes" src="https://user-images.githubusercontent.com/1204802/67269874-cdc5de00-f4b7-11e9-9b7c-ba73ddf2f6d2.png">

youknowriad commented:
What is the minimal version of this we can ship today that doesn't impact the non-FSE experience.

For me, it should be something like:

 - you just edit a post UI stays like today
 - enable FSE,  UI stays like today if you just edit a post
 - as soon as you edit something global, we can adapt the UI: Show the "Publish Changes"/"Update Changes" like the mockup above and leave the "save draft" button as is. (potentially in a v1, we can remove the "save draft" button entirely from the main UI since all this is still experimental). 

This also makes me thing we should hide the "Site Title" block behind the FSE flag for now (the other PR).

jasmussen commented:
> as soon as you edit something global, we can adapt the UI: Show the "Publish Changes"/"Update Changes" like the mockup above and leave the "save draft" button as is. (potentially in a v1, we can remove the "save draft" button entirely from the main UI since all this is still experimental).

Even just to test this, I would agree this is a good way to start. It let's us learn more about what the pain points are of saving, before optimizing too far in one direction. It's also quite close to this mockup: https://user-images.githubusercontent.com/1204802/67269874-cdc5de00-f4b7-11e9-9b7c-ba73ddf2f6d2.png

Copons commented:
I was trying to see how many different properties are handled in the Save Global Changes popup, and accidentally stumbled upon the Visibility selector.
Clicking on "Private" triggers a save and (privately) publish after accepting a popup.

<img width="812" alt="Screenshot 2019-10-22 at 15 12 29" src="https://user-images.githubusercontent.com/2070010/67295062-8bfc5e00-f4de-11e9-8f86-f64df60617c2.png">

In other words, just a warning that the Update/Publish button is not the only one that saves the post. 🙂

epiqueras commented:
>Maybe you're right — maybe we don't touch the save draft button, and all it does is save the post content. No new label indicating unsaved global changes, no extra buttons. But maybe we still change the label of the Publish button, and then show the popup. We might even show an "unread dot" or something, overlaid on the Publish button. Like this maybe?

Yep, that's what I think is optimal and that is what I meant by styling or iconizing the publish button when there are global changes. Some sort of "plus" or "more" icon comes to mind.

>For me, it should be something like:

That's how this PR works. I just think we should change it from rendering a new button to adding an icon to the publish button like discussed above.

>This also makes me thing we should hide the "Site Title" block behind the FSE flag for now (the other PR).

Yeah it is behind the flag.

jasmussen commented:
One thing I just remembered, which is a mistake in the mockups. Any "Publish" action that opens a modal dialog should have the "..." ellipsis at the end. So `● Publish Changes...` just like we have `Publish...`. The ellipsis implies it's the start of a flow.

shaunandrews commented:
I'm of the opinion that most of this should _just work_ and be handled "behind the scenes" without requiring users to make a decision or understand the entities and how they are related. Maybe I'm oversimplifying all of this, but I keep running through scenarios:

If I update the contents of a page, update the site title, and then I publish my changes they should all be saved and published.

If I update the contents of a page, update the site title, and then save the page as a draft the changes to the site title should also be saved as a draft.

If I update the site title, but never touch the contents of the page, and then publish my changes the site title should be updated.

If I update the site title, but never touch the contents of the page, and then schedule the page to publish, the changes to the site title should be saved and then published at the scheduled date.

--

Regardless of the above, I wonder why we're exploring a modal UI for showing/confirm changes to entities. We already have the pre-publish confirmation, and I think that would be a better place for showing all entities affected by saving/publishing/scheduling. Having it in the pre-publish confirmation would allow us to show the changes, but users could essentially ignore the various entities and just publish using the logic laid out above.

epiqueras commented:
For a lot of entities it doesn't make sense to draft or schedule. Those scenarios aren't really helpful to users and could create a lot of confusion.

What happens if I schedule different site title updates by scheduling two different posts? I assume the latest one will overwrite the previous one, but that is really confusing. We would need an entire API for this and a UI for warning and surfacing conflicts to users.

Furthermore, it becomes really hard to picture all the different things your save action will be affecting and users could unintentionally edit pages or template parts they didn't mean to.

This is also partly why the modal is decoupled from the pre publish flow, because these entities don't necessarily have a pre publish flow or any flow at all besides saving.

epiqueras commented:
@jasmussen @youknowriad 

>One thing I just remembered, which is a mistake in the mockups. Any "Publish" action that opens a modal dialog should have the "..." ellipsis at the end. So ● Publish Changes... just like we have Publish.... The ellipsis implies it's the start of a flow.

I just realized that this doesn't address the potentially very common use case of publishing global changes made while editing a draft you don't want to publish yet. This was the reason I opted for a separate button in the first place. Any ideas? We can always make publishing optional in the modal, but then we muddy the meaning of the button.

I think a separate button could work really well. In addition to the button we could have every block that saves outside the post have a special outline indicator to show that it has pending changes, and on hover we could give the option to save just that block.

The global changes save button would then become a quick way of saving all of these special blocks.

Have you thought about how we could build this as part of the block navigation toolbar. Just like the editor canvas, it could also highlight blocks like these and provide a "save all" button at the top.

Here's a doodle:

<img width="917" alt="Screen Shot 2019-10-22 at 5 37 09 PM" src="https://user-images.githubusercontent.com/19157096/67346282-0e8f1880-f4f3-11e9-97f7-0a950fc43f8a.png">

youknowriad commented:
> I just realized that this doesn't address the potentially very common use case of publishing global changes made while editing a draft you don't want to publish yet. This was the reason I opted for a separate button in the first place. Any ideas?

I still think a single button is the way to go. The user won't have to look into different places to find the save button for the thing he wants to save. I feel some of this complexity can be absorbed in the modal directly, maybe instead of having checkboxes per entity, we could have addition buttons next to each entity where it make sense. (For example a "save draft" button next to the post entity...)

jasmussen commented:
> I just realized that this doesn't address the potentially very common use case of publishing global changes made while editing a draft you don't want to publish yet. This was the reason I opted for a separate button in the first place. Any ideas?

I tried to address this in the mockup, by way of this bit:

<img width="448" alt="Screenshot 2019-10-23 at 09 03 31" src="https://user-images.githubusercontent.com/1204802/67366552-00351100-f574-11e9-892d-a1720782090c.png">

In this one, the `About Us` page which is in draft form is deliberately unchecked when you press the "Publish Changes" button. Which means when you press the Publish button, the only thing you publish is the change to the Header.

> I still think a single button is the way to go. The user won't have to look into different places to find the save button for the thing he wants to save. I feel some of this complexity can be absorbed in the modal directly, maybe instead of having checkboxes per entity, we could have addition buttons next to each entity where it make sense. (For example a "save draft" button next to the post entit

Strongly agree. Not just to keep this PR simple and minimal so we can actually test in master before we optimize down a path we're unsure of, but to keep things simple.

Copons commented:
> Regardless of the above, I wonder why we're exploring a modal UI for showing/confirm changes to entities. We already have the pre-publish confirmation, and I think that would be a better place for showing all entities affected by saving/publishing/scheduling. Having it in the pre-publish confirmation would allow us to show the changes, but users could essentially ignore the various entities and just publish using the logic laid out above.

@shaunandrews I think this is a very valid point.
If there are site-wide changes, we could force the pre-publish confirmation which could look something like this (disregard the very awkward copy of course):

<img width="558" alt="Screenshot 2019-10-23 at 11 39 14" src="https://user-images.githubusercontent.com/2070010/67384811-c53dd800-f589-11e9-9af5-c2c6acb1c97d.png">

epiqueras commented:
@jasmussen 

Good idea, labelling drafts and having them unchecked by default does the trick.

@Copons 

I also think that hooking into the pre publish flow like that is a better idea.

@jasmussen 

What do you think about going with that pre publish flow, and adding a button that says, "only publish site-wide changes", to accommodate for people publishing site-wide changes without the drafted post.

jasmussen commented:
> What do you think about going with that pre publish flow, and adding a button that says, "only publish site-wide changes", to accommodate for people publishing site-wide changes without the drafted post.

I think it's a pretty decent idea. I would love to see it explored slightly more in mockup form (@shaunandrews have the energy? I don't think we need to see much). 

The challenge is that the pre-publish flow is already pretty heavy. That's not to say it isn't the right place for it, and I still personally have a slight preference for the dialog. But the arguments presented hold promise.

youknowriad commented:
I think we have mockups to redesign the pre/post publish panels as modals/dialogs as well, this could be an opportunity to unify.

youknowriad commented:
I'd love to get a version of this landed (to be able to refactor the site title block), do you think we can implement the last suggestions here?

epiqueras commented:
We still don't have a clear decision here.

I think we should unify it with the pre publish flow. For that we also need a new state for the button to show that there are global edits on top of any post edits.

We need a mockup of how this fits in best. Some sort of icon for the button and a new panel in the pre-publish flow seems to be the clearest, shortest path forward.

@youknowriad @jasmussen

jasmussen commented:
> We still don't have a clear decision here.

Two paths forward, a panel in post-publish or a dialog. Contents of both should be the same, except perhaps the two Publish/Cancel buttons. Also, a changing label. This seems clear enough. Whether it's a panel or a dialog, I don't have a strong opinion on: is it more work to do one or the other if we wanted to test each direction?

> For that we also need a new state for the button to show that there are global edits on top of any post edits.

I don't think we'll want any new icons, unless you count the notification dot as an icon:

![Screenshot 2019-10-29 at 19 16 31](https://user-images.githubusercontent.com/1204802/67796620-a1462f00-fa80-11e9-88b1-ff6ac95292c9.png)

I think we should show that dot for when there are unpublished global changes other than the post/page itself.

This is the same pattern in many apps, it used to be there in MacOS: https://www.macobserver.com/tmo/article/mac_os_x_seeing_if_you_have_unsaved_changes

It's there in VS code:

![Screenshot 2019-10-29 at 19 18 37](https://user-images.githubusercontent.com/1204802/67796897-1c0f4a00-fa81-11e9-9a69-4334f809093e.png)

> We need a mockup of how this fits in best. Some sort of icon for the button and a new panel in the pre-publish flow seems to be the clearest, shortest path forward.

Okay here's a quick and dirty one:

![Screenshot 2019-10-29 at 19 23 22](https://user-images.githubusercontent.com/1204802/67797155-9e980980-fa81-11e9-8ae1-7128675b4f1e.png)

So actually this mockup makes me now lean towards the dialog, as opposed to the pre-publish flow. The pre-publish flow feels inherently tied to the _post itself_, and in this mockup and according to the flow suggested, global changes feel secondary. Additionally it feels like this makes it impossible to publish global changes on a draft, because it's a secondary action to publishing the post itself. Even more so because most of the stuff intended for the prepublish flow is for the post itself, so it becomes a weird mix of local and global.

There are discussions around making the prepublish flow a modal dialog itself. If this were to happen, the extra global step that happens could come before that, and lead in to the actual prepublish flow if you choose to also publish the post. Something like a wizard:

Flow A, user has only local changes:

1. User pressed "Publish..."
2. User receives prepublish flow in modal dialog, publishes, done.

Flow B, user has global changes in a draft but doesn't want to publish:

1. User makes global change, Publish... button changes to say `[dot] Publish Changes...`
2. User presses that button, receives global publish flow in modal dialog.
3. User chooses to only publish the global change, not the post. Pre-publish flow is not initiated.

Flow C, user has global changes and wants to publish the draft:

1. User makes global change after finishing their post. Presses button that says `[dot] Publish Changes...`.
2. User receives modal dialog with global publish flow. Checks to publish both post, and header change.
3. User presses Publish button. Snackbar says "Global Change published". Actual post pre-publish flow starts in a dialog (or sidebar until it becomes a dialog).
4. User publishes post.

epiqueras commented:
>I don't think we'll want any new icons, unless you count the notification dot as an icon:

Yes.

>Additionally it feels like this makes it impossible to publish global changes on a draft, because it's a secondary action to publishing the post itself. Even more so because most of the stuff intended for the prepublish flow is for the post itself, so it becomes a weird mix of local and global.

Agreed, we could add a button to only publish global changes, but it would still feel like a weird mix.

>There are discussions around making the prepublish flow a modal dialog itself. If this were to happen, the extra global step that happens could come before that, and lead in to the actual prepublish flow if you choose to also publish the post. Something like a wizard:

Yes, I think we should go with this. For now, it can lead to the current pre-publish flow, and in the future it can all happen in the same dialog if that's decided.

epiqueras commented:
Are we all happy with this direction? @mtias @youknowriad @jasmussen @Copons @shaunandrews

jasmussen commented:
> Are we all happy with this direction?

I think it's unlocks forward momentum without painting ourselves into a corner. In that vein, yes, I think this is a good step forward.

epiqueras committed:
Editor: Implement pre-pre-publish-flow global save flow.

epiqueras commented:
I am a bit sleepy and prepping for WCUS so I might have missed something, but I managed to squeeze this in tonight in the last commit.

I'm really liking this flow and I even found a Dashicon that might fit in well here.

Check it out:

## 1. Edited Post

<img width="889" alt="Screen Shot 2019-10-30 at 9 53 11 PM" src="https://user-images.githubusercontent.com/19157096/67920023-e9884e80-fb5f-11e9-8998-c0a5b3821e8a.png">

## 2. Edited Post & Site

<img width="882" alt="Screen Shot 2019-10-30 at 9 53 27 PM" src="https://user-images.githubusercontent.com/19157096/67920049-fc028800-fb5f-11e9-82ca-e941273e239b.png">

## 3. After Clicking Publish

<img width="884" alt="Screen Shot 2019-10-30 at 9 53 48 PM" src="https://user-images.githubusercontent.com/19157096/67920074-0d4b9480-fb60-11e9-9aa5-9b1c50009fc8.png">

## 4. After Clicking Save

<img width="882" alt="Screen Shot 2019-10-30 at 9 53 54 PM" src="https://user-images.githubusercontent.com/19157096/67920101-1f2d3780-fb60-11e9-803a-0e1a66d6aec8.png">

jasmussen commented:
I think this flow is good to ship because a) it solves a problem we have to solve in a relatively simple way, and b) it does not preclude us from future iterations.

However I do think before this branch can be merged, we need to use the circular dot instead of the dashicon — you can even draw a pseudo element and set border radius on it, doesn't have to be an icon. And can we change the label too?

epiqueras commented:
>I think this flow is good to ship because a) it solves a problem we have to solve in a relatively simple way, and b) it does not preclude us from future iterations.

Agreed.

>we need to use the circular dot instead of the dashicon

Can we do better than that? The dot would need at least a tooltip to differentiate it from regular changes.

>And can we change the label too?

To what? The mockups and discussion seem to leave it as is.

jasmussen commented:
> Can we do better than that? The dot would need at least a tooltip to differentiate it from regular changes.

The dot is in my opinion, a good baseline to start with, for the reasons outlined in [this comment](https://github.com/WordPress/gutenberg/pull/18029#issuecomment-547567040). We want to strike the balance between not overdoing it and not underdoing it. Always start with a minimal baseline and then iterate when it becomes clear that it's necessary to do so. Start simple.

If you include a tooltip, include it on the button as a whole, so we don't add any tab-stops. 

> To what? The mockups and discussion seem to leave it as is.

It's possible I'm missing context, but I'm still thinking [this flow](https://github.com/WordPress/gutenberg/pull/18029#issuecomment-547567040) makes sense:

- If you don't make any global changes, the big blue button says `Publish...`
- If you make a global change, the button says `[dot] Publish Changes...`

epiqueras committed:
Post Publish Button: Change dashicon to dot pseudo element.

epiqueras commented:
>The dot is in my opinion, a good baseline to start with, for the reasons outlined in this comment. We want to strike the balance between not overdoing it and not underdoing it. Always start with a minimal baseline and then iterate when it becomes clear that it's necessary to do so. Start simple.
>
>If you include a tooltip, include it on the button as a whole, so we don't add any tab-stops.

I think it clearly conveys that there are changes, but it doesn't communicate anything about the types of changes. It's something users will have to learn through trial: That the single dot means that there are global changes. I guess they already "learn" that `...` means that there is a pre publish flow so I guess the dot works for now. I've made the change.

>It's possible I'm missing context, but I'm still thinking this flow makes sense:
>
>- If you don't make any global changes, the big blue button says Publish...
>- If you make a global change, the button says [dot] Publish Changes...

`...`: There is a pre publish flow.
`[dot]`: There are global changes.

They are not mutually exclusive.

I think this is ready to test and merge now right?

jasmussen commented:
> I think it clearly conveys that there are changes, but it doesn't communicate anything about the types of changes

I will argue it doesn't have to. That is the task of the dialog that appears later, and/or the tooltip. 

> They are not mutually exclusive

They are not, but in the dialog flow we discussed, global changes come first, right?

> `...`: There is a pre publish flow.

Not exactly. The ellipsis just means "more to come", and that the effect of the button is not instant. Which is why we need it also for the "Publish Changes...":

![Screenshot 2019-11-01 at 09 45 00](https://user-images.githubusercontent.com/1204802/68013084-4d0f9a80-fc8c-11e9-9267-301424c25753.png)

If it just says "Publish Changes" it means the effect is instant.

epiqueras commented:
But, then we would be losing the difference between:

`Global Changes + Pre Publish Flow ([dot]...) vs. Global Changes ([dot])`

Doesn't `[dot]` already imply that it's not a regular instant publish?

jasmussen commented:
> But, then we would be losing the difference between:
> Global Changes + Pre Publish Flow ([dot]...) vs. Global Changes ([dot])

I don't think that distinction is valuable or even necessary, to be honest.

> Doesn't [dot] already imply that it's not a regular instant publish?

I see what you're saying, but I don't think that's a reason to not include the ellipsis.

If you feel strongly about this, we could get more designers than me looking at this?

epiqueras committed:
Post Publish Button: Also add trailing ellipsis when there are global changes.

epiqueras commented:
>If you feel strongly about this, we could get more designers than me looking at this?

Not really, I just wanted to make sure you had thought about it. I've made the change :smile:

jasmussen commented:
> Not really, I just wanted to make sure you had thought about it. I've made the change 😄

Haha sometimes these things wake me up in the middle of the night!

I think what's important with many of the decisions is evaluating whether the change is _difficult to change_ or not. If it's trivial to change, then there's no reason not to just go with the most obvious solution and revisit if need be; move forward and test it in the plugin. I'm sure this interface will be iterated many times. It might even end up in the pre publish flow despite our current instincts.

epiqueras commented:
Haha me too.

Yeah exactly, and this is a pretty trivial change as you can see from the
commit, haha.

On Fri, Nov 1, 2019 at 9:46 AM Joen Asmussen <notifications@github.com>
wrote:

> Not really, I just wanted to make sure you had thought about it. I've made
> the change 😄
>
> Haha sometimes these things wake me up in the middle of the night!
>
> I think what's important with many of the decisions is evaluating whether
> the change is *difficult to change* or not. If it's trivial to change,
> then there's no reason not to just go with the most obvious solution and
> revisit if need be; move forward and test it in the plugin. I'm sure this
> interface will be iterated many times. It might even end up in the pre
> publish flow despite our current instincts.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/18029?email_source=notifications&email_token=AESFA2ALWSM3ZVOWZHPSOALQRQ6L3A5CNFSM4JCL7362YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEC3D7UI#issuecomment-548814801>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESFA2EXZ3YATHP2YQZJROTQRQ6L3ANCNFSM4JCL736Q>
> .
>

youknowriad reviewed (commented):
Some bugs I noticed while testing this:

 - Clicking "Publish" when you have multiple changes and you select everything and save, triggers the "pre publish" panel. Ideally, these two steps (pre publish panel and the modal) are just one unique step like discussed before but for now, I think we should skip the pre publish panel entirely if we use the modal. But It's fine to show the post publish panel.

 - If I have multiple changes and my post is published, it says "publish" instead of "update"

youknowriad commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,98 @@
+} ) => (
+	<CheckboxControl
+		label={ `${ startCase( name ) }: "${ rawRecord.title || rawRecord.name }"` }
+		help={ `Changed Properties: ${ Object.keys( edits ).join( ', ' ) }.` }
```
I think that part is not really useful and just confusing. I'd suggest just removing it. (the changed properties helper)

epiqueras replied:
Agreed, at least until we have a better UI for it.

epiqueras committed:
Entities Saved States: Remove changed properties text.

epiqueras committed:
Entities Saved States: Use the correct label.

epiqueras commented:
>Clicking "Publish" when you have multiple changes and you select everything and save, triggers the "pre publish" panel. Ideally, these two steps (pre publish panel and the modal) are just one unique step like discussed before but for now, I think we should skip the pre publish panel entirely if we use the modal. But It's fine to show the post publish panel.

This is what we agreed to:

>>So actually this mockup makes me now lean towards the dialog, as opposed to the pre-publish flow. The pre-publish flow feels inherently tied to the post itself, and in this mockup and according to the flow suggested, global changes feel secondary. Additionally it feels like this makes it impossible to publish global changes on a draft, because it's a secondary action to publishing the post itself. Even more so because most of the stuff intended for the prepublish flow is for the post itself, so it becomes a weird mix of local and global.
>>
>>There are discussions around making the prepublish flow a modal dialog itself. If this were to happen, the extra global step that happens could come before that, and lead in to the actual prepublish flow if you choose to also publish the post. Something like a wizard:...

>If I have multiple changes and my post is published, it says "publish" instead of "update"

Fixed!

youknowriad commented:
> This is what we agreed to:

We agreed on using the dialog yes, but not the dialog and the prepublish panel at the same time. I'm thinking we should skip the prepublish panel when the dialog is used.

epiqueras commented:
From @jasmussen:

>If this were to happen, the extra global step that happens could come before that, and lead in to the actual prepublish flow if you choose to also publish the post.

We should not take away the features of the pre publish flow from users.

youknowriad commented:
> We should not take away the features of the pre publish flow from users.

Since this is a new flow only enabled in FSE (experimental) when you have multiple entities, I don't think we're taking away any feature. So for me it's just confusing, I didn't undestand why the post was not saved and the prepublish panel opened, I wasn't even certain when the "site" saving happened, was it before the prepublish panel or after, it's really very confusing.

@jasmussen Thoughts on this?

epiqueras commented:
>I wasn't even certain when the "site" saving happened, was it before the prepublish panel or after

We need a notice for that.

youknowriad commented:
> We need a notice for that.

A notice is definitely needed, we do have one on post publish already, I just think save should happen at the same time for everything and we can just leave the button of the modal in a loading state while saving.

jasmussen commented:
> Since this is a new flow only enabled in FSE (experimental) when you have multiple entities, I don't think we're taking away any feature. So for me it's just confusing, I didn't undestand why the post was not saved and the prepublish panel opened, I wasn't even certain when the "site" saving happened, was it before the prepublish panel or after, it's really very confusing.
> @jasmussen Thoughts on this?

So to clarify my thinking it's important to distinguish between the act of _saving_ and the act of _publishing_.

My thinking is that if, as part of the popup dialog, you _publish_ the post or page you're editing, I do feel strongly that the pre-publish flow should come after the dialog. Reason being that the editor was designed to enable a user to dismiss the sidebar and never ever open it again if they didn't want to. The prepublish flow exists for that reason, to give a last minute chance to set tags, categories, publish dates and so on. So the pre-publish flow matters _only_ when a post or page or other cpt is _published_ along with any other global changes. 

While I do feel strongly about this (and I believe a dialog version of the pre-publish flow could make this less jarring), I also don't want _perfect_ to stand in the way of _good_, especially since we're in the plugin phase. So if we want to get this in to test it, get a feeler for it, and then intend to continue iterating on it, with the above in mind, all good on my front then.

epiqueras commented:
>While I do feel strongly about this (and I believe a dialog version of the pre-publish flow could make this less jarring), I also don't want perfect to stand in the way of good, especially since we're in the plugin phase. So if we want to get this in to test it, get a feeler for it, and then intend to continue iterating on it, with the above in mind, all good on my front then.

This is how it is implemented. @youknowriad was suggesting we skip the pre publish flow altogether after the popup dialog. I also think we should keep it for the same reasons you laid out.

epiqueras commented:
@youknowriad can we merge this now?

youknowriad commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,97 @@
+);
+
+export default function EntitiesSavedStates( { isOpen, onRequestClose } ) {
+	const entityRecordChangesByRecord = useSelect( ( select ) =>
```
If I understand properly, we will be receiving changes from all entities here even entities that are not edited in the editor itself (think plugin, left over from other pages if it's an SPA...).

Do you think the "editor" should register the entities it's editing somehow instead of just getting all edits from the core data store?

epiqueras replied:
>If I understand properly, we will be receiving changes from all entities here even entities that are not edited in the editor itself (think plugin, left over from other pages if it's an SPA...).

Yes, I thought that wouldn't be a blocker since we only have one editor for now.

>Do you think the "editor" should register the entities it's editing somehow instead of just getting all edits from the core data store?

Yes, I think the `EntityProvider` should trigger an action that adds the provided entity record to a list in the editor store on mount. The editor can then use this list to query/filter entity changes.

Should I do that in this PR?

youknowriad replied:
I'm not sure it's the responsibility of `EntityProvider ` to do that as EntityProvider is not editor specific either.

I'm ok for this being in a separate PR.

epiqueras replied:
No, but it is meant to be rendered in an editing context and it's the only way to link the two dynamically without walking the block tree.

I.e. The only way for the editor to answer: Which entities am I rendering/editing?

youknowriad replied:
I was imagining something like the EditorProvider registering its entities and maybe the Post block, Site block registering their own. I can see how it's not great either as extra work for block authors. We can try and iterate.

epiqueras replied:
Yeah, and we can do it in `EntityProvider` in a way that does not break usage outside of an editor.

shaunandrews commented:
Curious if we could expose the changes to be published in the pre-publish confirmation sidebar — rather than just in modal:

<img width="275" alt="image" src="https://user-images.githubusercontent.com/191598/69664108-b1d4cf80-1055-11ea-97af-5a054a27c1a2.png">

youknowriad commented:
@shaunandrews Great proposal, I like it. Right now that panel don't show up for updates and it can be completely dismissed as well in a setting. Do you think we should force the panel to show when we have multiple entities being edited?

epiqueras commented:
We discussed this approach before in this PR.

I agree with @jasmussen on this:

>So actually this mockup makes me now lean towards the dialog, as opposed to the pre-publish flow. The pre-publish flow feels inherently tied to the post itself, and in this mockup and according to the flow suggested, global changes feel secondary. Additionally it feels like this makes it impossible to publish global changes on a draft, because it's a secondary action to publishing the post itself. Even more so because most of the stuff intended for the prepublish flow is for the post itself, so it becomes a weird mix of local and global.
>
>There are discussions around making the prepublish flow a modal dialog itself. If this were to happen, the extra global step that happens could come before that, and lead in to the actual prepublish flow if you choose to also publish the post. Something like a wizard:

As a user, it makes me raise questions like, does the visibility setting apply to all my changes? What about scheduling?

I also think this is something that can be explored as we go along, specially if there are discussions around making the pre publish flow a modal dialog itself, and we should not block this PR.

mtias commented:
I agree as well, but only in the context of a post. In the context of an `edit-site` I don't think this remains true:

> So actually this mockup makes me now lean towards the dialog, as opposed to the pre-publish flow. The pre-publish flow feels inherently tied to the post itself, and in this mockup and according to the flow suggested, global changes feel secondary.

epiqueras commented:
>In the context of an edit-site I don't think this remains true:

In the context of an `edit-site`, what even is in the pre-publish flow? Few, if any, of the same settings for posts apply. The sidebar would likely just contain the global changes saving section and wouldn't be confusing. The issue is mixing it with the post's pre-publish flow.

So, if we want to keep this in the sidebar, I would suggest having a pre-pre-publish flow for the global changes that then leads to the pre-publish flow if the post was selected/checked. If we are in `edit-site` where there is no "post", then we would just have the pre-pre-publish flow.

But, if we are going to move all of this to a modal dialog later anyways, it doesn't make sense to invest in it.

mtias commented:
> The issue is mixing it with the post's pre-publish flow.

Pre-publish should be able to display this if enabled. It might not be the default way to present things, but it's still appropriate for someone to want to show it there.

youknowriad commented on packages/core-data/src/selectors.js:
```diff
@@ -153,6 +153,52 @@ export function getEntityRecords( state, kind, name, query ) {
+ *
+ * @return {Array} The list of edited records with their edits.
+ */
+export const getEntityRecordChangesByRecord = createSelector(
```
I'm not sure we'll need such a selector. Each time the returned value from a selector is that complex, I feel something is wrong. Can we just replace it with a selector that checks if an entity is Dirty?
I guess right now we'd also need a selector to retrieve all dirty records (hopefully we could remove that later).

epiqueras replied:
>I guess right now we'd also need a selector to retrieve all dirty records (hopefully we could remove that later).

Yes we would need that anyways.

youknowriad commented:
It feels like we're down a rabbit hole a little bit here. Should we just go with what's in the PR and iterate? I personally don't like at all the double steps after clicking the "Publish" button, it's too disturbing as a user but I'm fine moving forward and iterating as this PR is an important piece in order to remove save buttons from the entity blocks.

epiqueras commented:
Agreed, but I can't approve my own PR :laughing:

aduth commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,97 @@
+					onClick={ saveCheckedEntities }
+					className="editor-entities-saved-states__save-button"
+				>
+					Save
```
Should be localized.

aduth commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -102,20 +140,32 @@ export class PostPublishButton extends Component {
 					ref={ this.buttonNode }
 					{ ...componentProps }
+					className={ classnames( componentProps.className, {
+						'editor-post-publish-button__has-changes-dot': hasNonPostEntityChanges,
```
This reads like it should be a modifier class to me:

```suggestion
						'has-changes-dot': hasNonPostEntityChanges,
```

Which could then be referenced as:

```scss
.editor-post-publish-button .components-button.has-changes-dot { /* ... */ }
```

Or as a modifier on the publish button itself:

```scss
.editor-post-publish-button.has-changes-dot .components-button { /* ... */ }
```

Or at the very least as a modifier of a class specific to PostPublishButton, assigned to the button:

```suggestion
						'editor-post-publish-button__button': true,
						'has-changes-dot': hasNonPostEntityChanges,
```

(instead of `true` value assignment, can be a separate argument of `classnames`, outside the object)

```scss
.editor-post-publish-button__button.has-changes-dot .components-button { /* ... */ }
```

Reference: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#naming

epiqueras replied:
>Or at the very least as a modifier of a class specific to PostPublishButton, assigned to the button:

This makes the most sense.

It's too specific to be a modifier of `.components-button` and it's not modifying all of `.editor-post-publish-button`.

aduth commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,97 @@
+	);
+	const { saveEditedEntityRecord } = useDispatch( 'core' );
+
+	const [ checkedById, _setCheckedById ] = useState( {} );
```
Not really a blocker, and maybe discussed previously, but wouldn't it be more expected that saving is an opt-out, not an opt-in?

epiqueras replied:
I think we should only check the current post by default, but we could leave that for another PR. There was also discussion around how drafts should/should not be checked.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,97 @@
+	setCheckedById,
+} ) => (
+	<CheckboxControl
+		label={ `${ startCase( name ) }: "${ rawRecord.title || rawRecord.name }"` }
```
It shows the saved title/name value, not the edited one.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,97 @@
+	<CheckboxControl
+		label={ `${ startCase( name ) }: "${ rawRecord.title || rawRecord.name }"` }
+		checked={ checked }
+		onChange={ useCallback( ( nextChecked ) => setCheckedById( id, nextChecked ), [ id ] ) }
```
>It might be expected that setCheckedById should be listed as a dependency

It never changes. It's like an extension of the function returned by `useState` here.

>But to the more general application, I don't know that we really care to use useCallback here:

Yeah, none of the components here are expensive and it's a leaf component.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,97 @@
+	);
+	const saveCheckedEntities = useCallback( () => {
+		Object.keys( checkedById ).forEach( ( id ) =>
+			saveEditedEntityRecord( ...id.split( ' | ' ).filter( ( s ) => s !== 'undefined' ) )
```
I was hopeful that `undefined` would never be used.

I'll change it to `EquivalentKeyMap` to be sound.

epiqueras replied:
To clarify, we should use `EquivalentKeyMap`, because `kind` or `name` might contain the string `"undefined"` or `" | "` and break the logic. But, we still need to check and filter out the `id`/`key` if it equals the string `"undefined"`, because that's how base entities like root-site are stored in the store.

aduth replied:
> But, we still need to check and filter out the `id`/`key` if it equals the string `"undefined"`, because that's how base entities like root-site are stored in the store.

I guess it's not something we need to solve here, but that doesn't sound like an ideal way to structure the state. More broadly, I feel like the entire "kind", "name" pair system is pretty clunky in its usage, and would be worth revisiting.

Note: I'm just basing this off observation alone; I don't really know how viable (or the full extent of what would be involved in) refactoring this state.

epiqueras replied:
Agreed, unfortunately, this was already very ingrained in all of `core-data`.

We would need to move undefined-ID entities into their own slice of state instead of being nested inside `kind => name`. There would be a lot of changes involved.

I think it's safe to assume/enforce that IDs won't/don't `=== 'undefined'`.

epiqueras commented on packages/core-data/src/selectors.js:
```diff
@@ -153,6 +153,52 @@ export function getEntityRecords( state, kind, name, query ) {
+ *
+ * @param {Object} state State tree.
+ *
+ * @return {Array} The list of edited records with their edits.
```
Yeah, this was a mistake. Fixed.

I've also made the description clearer.

epiqueras committed:
Post Publish Button: Review suggestions.

epiqueras committed:
Package: Update lockfile.

epiqueras commented:
@aduth thanks for the review! All done :smile:

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -134,6 +134,65 @@ export const isEditedPostDirty = createRegistrySelector( ( select ) => ( state )
+ *
+ * @return {boolean} Whether there are edits or not.
+ */
+export const hasNonPostEntityChanges = createRegistrySelector(
```
It would be good to have some unit tests for this selector. And also the one in the other file (`getEntityRecordChangesByRecord`).

aduth reviewed (commented):
I seem to be having some trouble with saving, when edits include those to a block and to a separate entity.

1. Navigate to Posts > Add New
2. Provide a title to the post
3. Publish the post
4. Add a Site Title block
5. Edit the site title in the Site Title block
6. Click "Publish..."
7. Select both options, to save both the post and the site title
8. Press "Save"
9. Reload the page
10. Note that you are not prompted that there are unsaved changes (indicating that the editor considers everything to have been saved), but when the page finishes reloading, the block added in Step 4 is absent.

epiqueras committed:
Core Data: Improve `getEntityRecordChangesByRecord` types and docs.

epiqueras committed:
Entities Saved States: Add inline comments and fix post saving race condition.

epiqueras committed:
Core Data: Test `getEntityRecordChangesByRecord`.

epiqueras committed:
Editor: Test `hasNonPostEntityChanges`.

epiqueras commented on packages/core-data/src/selectors.js:
```diff
@@ -153,6 +153,54 @@ export function getEntityRecords( state, kind, name, query ) {
+ *
+ * @param {Object} state State tree.
+ *
+ * @return {Object} The map of edited records with their edits.
```
Updated

Note that this breaks JSDoc doc generation parameter detection.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,98 @@
+	);
+	const { saveEditedEntityRecord } = useDispatch( 'core' );
+
+	const [ checkedById, _setCheckedById ] = useState( new EquivalentKeyMap() );
```
I think the overhead of a new function is probably the same, but I added it for convention.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,98 @@
+		onRequestClose( checkedById );
+	};
+	return (
+		isOpen && (
```
This is actually done by prettier.

epiqueras commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,98 @@
+						( changedName ) =>
+							Object.keys(
+								entityRecordChangesByRecord[ changedKind ][ changedName ]
+							).map( ( changedKey ) => {
```
I doubt their flatMap implementation is less performant.

epiqueras commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -15,18 +16,50 @@ import { __ } from '@wordpress/i18n';
+				entitiesSavedStatesCallback();
+			}
+		} );
+	}
```
Made it clearer.

epiqueras commented:
>I seem to be having some trouble with saving, when edits include those to a block and to a separate entity.

There was a race condition with the modal setting `isSaving` to true and then making the held callback's dispatched `savePost` bail early. I fixed it.

`savePost` makes sure the latest serialized block content is in `core-data` while the modal's save hook doesn't, which is why it missed the most recent block.

I also implemented all other suggestions and wrote tests for the new selectors.

aduth commented on packages/editor/src/components/entities-saved-states/index.js:
```diff
@@ -0,0 +1,104 @@
+		isOpen && (
+			<Modal
+				title={ __( 'What do you want to save?' ) }
+				onRequestClose={ onRequestClose }
```
This won't pass an argument to `closeEntitiesSavedStates`, which causes an error:

```
TypeError: savedById.has is not a function
```

https://github.com/WordPress/gutenberg/blob/ff5da59/packages/editor/src/components/post-publish-button/index.js#L68

Steps:

1. Update Site Title block and post title
2. Press "Publish..."
3. Press <kbd>Escape</kbd>

epiqueras replied:
Fixed.

epiqueras committed:
Entities Saved States: Fix cancel close error.

aduth commented on packages/editor/src/components/post-publish-button/index.js:
```diff
@@ -45,7 +90,14 @@ export class PostPublishButton extends Component {
+			entitiesSavedStatesCallback,
+		} = this.state;
+
 		const isButtonDisabled =
```
I think this needs to be accounting for `hasNonPostEntityChanges`.

Otherwise, if I have a post which includes _only_ changes to a non-post entity, it still appears as disabled:

![image](https://user-images.githubusercontent.com/1779930/70754412-2de13f80-1d05-11ea-954c-74d6d6d91b73.png)

But I can still click on it to show the save prompt:

![image](https://user-images.githubusercontent.com/1779930/70754447-3b96c500-1d05-11ea-964d-5472f38f4491.png)

epiqueras replied:
Fixed.

epiqueras committed:
Post Publish Button: Enable buttons when there are non post entity changes.

youknowriad commented:
Thanks everyone for the reviews and work here. 

I guess the next step here would be to consolidate these pre/post publish flows. It would be good to have some design direction/exploration/mockups.
